### DISCUSSION TITLE
Cancel FSI interactions via FCS net7?

### DISCUSSION THREAD

**Discussion by goswinr** [dotnet/fsharp#14491] [Q&A]
How to cancel FSI interactions in net7 ?
The new internal [ControlledExecution.fs](https://github.com/dotnet/fsharp/blob/main/src/Compiler/Interactive/ControlledExecution.fs) class seems to work well in VS for cancelling.
But I have not found a public member that exposes it to the FCS.
The cancelation token passed in via `fsiSession.EvalInteraction(code,token)` only seems to reach till here:  https://github.com/dotnet/fsharp/blob/ad65b819fc2c5b6b1beae7b3873da550f8a14af8/src/Compiler/Interactive/fsi.fs#L2937
I could not use it successfully.

Does it reach all the way to ControlledExecution.fs ? This class seems to have its own token.  TryAbort: https://github.com/dotnet/fsharp/blob/ad65b819fc2c5b6b1beae7b3873da550f8a14af8/src/Compiler/Interactive/ControlledExecution.fs#L49

I also tried to run my own [ControlledExecution Class](https://learn.microsoft.com/en-us/dotNet/API/system.runtime.controlledexecution?view=net-7.0) 
but got the message `Compiler Error:input.fsx (1,1)-(1,1) interactive error internal error: The thread is already executing the ControlledExecution.Run method.` 
Did I miss something ? Or is the `token` in the `Eval` method suposed to work and I might be using it wrongly ?

**Comment by vzarytovskii** [comment]
@KevinRansom what would be the proper way of doing that?

**Reply by goswinr** [reply]
@KevinRansom @vzarytovskii would you accept this commit as pull request to enable cancellation? https://github.com/goswinr/fsharp/commit/87f653405bde44c779e78d95cda80b80a6bceb58

**Reply by T-Gro** [reply]
How would that work in this scenario, i.e. how would we ensure (instrument?) to make sure the token is being checked and execution stopped?

Or do I imagine it wrongly that user code would continue to be running despite of this, unless the token is explicitly checked?


/// On net7.0 or later it uses **a cancellation token** in System.Runtime.ControlledExecution.Run()

**Reply by goswinr** [reply]
@T-Gro I am not sure if I understand your question. The token is internal here https://github.com/dotnet/fsharp/blob/e30d14cb46f290050ac8e2bbea5e9b804b97bdde/src/Compiler/Interactive/ControlledExecution.fs#L21 that is why I would like to make ‘TryAbort’ public. The build from my fork works for my use case. (cancelling an FSI session hosted in an WPF app)

**Comment by goswinr** [comment]
I created a pull request to enable this https://github.com/dotnet/fsharp/pull/14546

**Comment by goswinr** [comment]
solved by https://github.com/dotnet/fsharp/pull/15184
